sql/src/main/java/io/crate/planner/operators/NestedLoopJoin.java

```diff
@@ -284,6 +284,42 @@ public LogicalPlan pruneOutputsExcept(Collection<Symbol> outputsToKeep) {
         );
     }
 
+    @Nullable
+    @Override
+    public FetchRewrite rewriteToFetch(Collection<Symbol> usedColumns) {
+        ArrayList<Symbol> usedFromLeft = new ArrayList<>();
+        ArrayList<Symbol> usedFromRight = new ArrayList<>();
+        for (Symbol usedColumn : usedColumns) {
+            SymbolVisitors.intersection(usedColumn, lhs.outputs(), usedFromLeft::add);
+            SymbolVisitors.intersection(usedColumn, rhs.outputs(), usedFromRight::add);
+        }
+        if (joinCondition != null) {
+            SymbolVisitors.intersection(joinCondition, lhs.outputs(), usedFromLeft::add);
+            SymbolVisitors.intersection(joinCondition, rhs.outputs(), usedFromRight::add);
+        }
+        FetchRewrite lhsFetchRewrite = lhs.rewriteToFetch(usedFromLeft);
+        if (lhsFetchRewrite == null) {
+            return null;
+        }
+        FetchRewrite rhsFetchRewrite = rhs.rewriteToFetch(usedFromRight);
+        if (rhsFetchRewrite == null) {
+            return null;
+        }
+        return new FetchRewrite(
+            Maps.concat(lhsFetchRewrite.replacedOutputs(), rhsFetchRewrite.replacedOutputs()),
+            new NestedLoopJoin(
+                lhsFetchRewrite.newPlan(),
+                rhsFetchRewrite.newPlan(),
+                joinType,
+                joinCondition,
+                isFiltered,
+                topMostLeftRelation,
+                orderByWasPushedDown,
+                rewriteFilterOnOuterJoinToInnerJoinDone
+            )
+        );
```

Comment:
Looks like there is no LogicalPlan test for this optimization, maybe worth creating one?